app/image/gc/gc.go

```diff
@@ -67,111 +123,331 @@ func (g *imgGC) spin() {
+		a, err := app.GetByName(appVersions.AppName)
 		if err != nil && err != appTypes.ErrAppNotFound {
 			multi.Add(err)
 			continue
 		}
 		if a == nil {
-			log.Debugf("[image gc] app %q not found, removing everything", versions.AppName)
-			err = registry.RemoveAppImages(versions.AppName)
-			if err != nil {
-				multi.Add(err)
-			}
-			err = servicemanager.AppVersion.DeleteVersions(versions.AppName)
+			log.Debugf("[image gc] app %q not found, removing everything", appVersions.AppName)
+			err = pruneAllVersionsByApp(appVersions.AppName)
 			if err != nil {
 				multi.Add(err)
 			}
 			continue
 		}
 
-		var regularVersions, customTagVersions []appTypes.AppVersionInfo
-		for _, v := range versions.Versions {
-			if v.CustomBuildTag != "" {
-				customTagVersions = append(customTagVersions, v)
-			} else {
-				regularVersions = append(regularVersions, v)
-			}
+		//
+		requireExclusiveLock, err := markOldImagesForAppVersion(a, appVersions, historySize, false)
+		if err != nil {
+			multi.Add(err)
+			continue
+		}
+		if !requireExclusiveLock {
+			continue
 		}
 
-		sort.Slice(regularVersions, func(i, j int) bool {
-			return regularVersions[i].UpdatedAt.After(regularVersions[j].UpdatedAt)
+		evt, err := event.NewInternal(&event.Opts{
+			Target:       event.Target{Type: event.TargetTypeApp, Value: appVersions.AppName},
+			InternalKind: "version gc",
+			Allowed:      event.Allowed(permission.PermAppReadEvents, permission.Context(permTypes.CtxApp, appVersions.AppName)),
 		})
 
-		sort.Slice(customTagVersions, func(i, j int) bool {
-			return customTagVersions[i].UpdatedAt.After(customTagVersions[j].UpdatedAt)
-		})
+		if err != nil {
+			if _, ok := err.(event.ErrEventLocked); !ok {
+				continue
+			}
+			multi.Add(errors.Wrapf(err, "unable to acquire lock of app: %q", appVersions.AppName))
+			continue
+		}
+
+		_, err = markOldImagesForAppVersion(a, appVersions, historySize, true)
+		if err != nil {
+			multi.Add(err)
+		}
+		evt.Done(err)
+	}
+	return multi.ToError()
+}
+
+func markOldImagesForAppVersion(a *app.App, appVersions appTypes.AppVersions, historySize int, exclusiveLockAcquired bool) (requireExclusiveLock bool, err error) {
+	deployedVersions, err := a.DeployedVersions()
+	if err == app.ErrNoVersionProvisioner {
+		deployedVersions = []int{appVersions.LastSuccessfulVersion}
+	} else if err != nil {
+		return false, errors.Wrapf(err, "Could not get deployed versions of app: %s", appVersions.AppName)
+	}
+
+	versionsToRemove, versionsToPruneFromProvisioner := selectAppVersions(appVersions, deployedVersions, historySize)
+	if !exclusiveLockAcquired {
+		for _, version := range versionsToPruneFromProvisioner {
+			pruneVersionFromProvisioner(a, version)
+		}
+	}
+
+	if len(versionsToRemove) == 0 {
+		return false, nil
+	} else if !exclusiveLockAcquired {
+		return true, nil
+	}
+
+	for _, version := range versionsToRemove {
+		versionsMarkedToRemovalTotal.Inc()
+		pruned := pruneVersionFromProvisioner(a, version)
```

Comment:
What do you think of moving this to the sweep step? The reasoning would be to keep the lock for as little time as possible and this call could potentially take a while. The previous call for pruneVersionFromProvisioner in this function is fine because we're not locking yet.